### DISCUSSION TITLE
Mender Add-on Filetransfer (Troubleshoot)

### DISCUSSION THREAD

**Discussion by pasco1995** [joelguittet/mender-mcu-client#79] [General]
Hi @joelguittet,

I wanted to ask if you are already working on new mender features like the file transfer ?

I am currently working on a esp32 project where I would like to use this feature.

Is there any mender documentation on the protocol, how it has to be set up on the client side?

**Comment by joelguittet** [comment]
Hello @pasco1995 

This is a feature I would like to add for sure, but not done a lot of work on it for the moment (except some error logs that should be displayed indicating it is not supported in the client).
There is no documentation at all on the troubleshoot add-on protocol from Mender. I have done the remote console part by reverse engineering the code.

I'm really opened to this kind of work if some people are interested. Let me few days to check the current status and if I can provide at least a draft of this. I will be in vacation end of next week so I don't promise anything. If not finished I can finalize in September.

Does it sounds good for you ?

Joel

**Comment by joelguittet** [comment]
Follow up of this activity here: https://github.com/joelguittet/mender-mcu-client/issues/80

**Comment by pasco1995** [comment]
Hi @joelguittet.

that sounds great ! Thanks

**Comment by joelguittet** [comment]
@pasco1995 a draft version of the mender-esp32-example supporting file transfer is available at https://github.com/joelguittet/mender-esp32-example/tree/feature/add-on-troubleshoot-file-transfer. Take care to update the submodule too. Then you can build and flash as usual.

I have added a litllefs partition to upload/download files and tested with fiels up to 100KB with no issues actually. littlefs is just for the example, you can use nvs or any other media. Just need to write your own file transfer callbacks accordingly of course.

Important note: the troubleshoot add-on APIs have changed so if you currently work in your own project, please look at the callbacks initialization in the example and update accordingly (the commit on the mender-esp32-example is perfect to see the impacts).

I will really appreciate feedbacks from you on this feature has it is very new and forced me to reworked totally the troubleshoot addon architecture to integrate this. Thanks!

Note I will finalize the support for Zephyr platform soon as well (September).

**Comment by joelguittet** [comment]
Note: also a memory leak fixed in the shell callback in the example 👀

**Comment by pasco1995** [comment]
Hi @joelguittet 

Really cool, thanks a lot!
I will try it next week and give you feedback.

**Reply by joelguittet** [reply]
Great. 
I had a look to the zephyr support as well but for the moment not working, but I guess you are using esp-idf.

**Reply by pasco1995** [reply]
Right

**Comment by joelguittet** [comment]
Net step for this feature to be followed in https://github.com/joelguittet/mender-mcu-client/pull/81
